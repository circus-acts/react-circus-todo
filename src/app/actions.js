// The actions defined here simply push event values into circuit signals.

import circuit from './circuit'
const {filterBy, editing, todos} = circuit.signals

// Create a helper function to get event value by type and push it into a signal.
const fromEvent = (signal, id) => ({keyCode, type, target}) => {
  if (keyCode === 13 || type === 'blur' || target.type === 'checkbox') {
    const value = target[target.type === 'checkbox' ? 'checked' : 'value']
    if (id === 'add') {
      signal(value)
      target.value = ''
    }
    else signal({id, value})
  }
}

const actions = {
  add: fromEvent(todos.add, 'add'),
  // the following signals are wrapped in a binding identity
  bind: id => ({
    update: fromEvent(todos.update, id),
    complete: fromEvent(todos.complete, id),
    remove: () => todos.remove(id),
    edit: () => editing(id)
  }),

  // pass these signals through unchanged - just for convenience really
  purge: todos.purge,
  toggle: todos.toggle,
  ALL: filterBy.ALL,
  ACTIVE: filterBy.ACTIVE,
  COMPLETED: filterBy.COMPLETED
}

export default actions
