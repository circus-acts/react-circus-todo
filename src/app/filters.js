//Export some constants to share through the app.
export const ALL = 'ALL'
export const ACTIVE = 'ACTIVE'
export const COMPLETED = 'COMPLETED'

// This function will be mapped over the circuit output so its signature
// is going to match the values provided by the primary circuit channels.

export default channels => {
  const {todos = [], filterBy} = channels
  return {
    ...channels,
    todos: todos.filter(todo => {
      switch (filterBy) {
        case COMPLETED: return todo.completed
        case ACTIVE: return !todo.completed
        default: return true
      }
    }),
    total: todos.length,
    remaining: todos.reduce((r, todo) => r + (todo.completed? 0 : 1), 0)
  }
}
