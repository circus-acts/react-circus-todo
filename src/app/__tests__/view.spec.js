import React from 'react'
import renderer from 'react-test-renderer'
import View, {Todo, Filter} from '../view'
import actions from '../actions'
import {ALL, ACTIVE, COMPLETED} from '../filters'

describe('view', () => {

    it('should render a todo', () => {
        const data = {
            description: 'item 1',
            completed: false
        }
        const component = renderer.create(
            <Todo data={data} actions={actions}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should render a todo for editing', () => {
        const data = {
            description: 'item 1',
            completed: false
        }
        const component = renderer.create(
            <Todo editing data={data} actions={actions}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should render a selected filter', () => {
        const component = renderer.create(
            <Filter by={ALL} filterBy={ALL}/>
        );
        let tree = component.toJSON();
        expect(tree.children[0].props.className).toBe('selected')
        expect(tree).toMatchSnapshot();
    })

    it('should render an un-selected filter', () => {
        const component = renderer.create(
            <Filter by={ALL} filterBy={ACTIVE}/>
        );
        let tree = component.toJSON();
        expect(tree.children[0].props.className).toBe('')
        expect(tree).toMatchSnapshot();
    })

    it('should render a view', () => {
        const data = {
            todos: []
        }
        const component = renderer.create(
            <View {...data}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })

})
