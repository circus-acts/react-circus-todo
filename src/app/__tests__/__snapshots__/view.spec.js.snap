exports[`view should render a selected filter 1`] = `
<li>
  <a
    className="selected"
    onClick={[Function]}>
    All
  </a>
</li>
`;

exports[`view should render a todo 1`] = `
<li
  className={false}>
  <div
    className="view">
    <input
      checked={false}
      className="toggle"
      onChange={undefined}
      type="checkbox" />
    <label
      onDoubleClick={undefined}>
      item 1
    </label>
    <button
      className="destroy"
      onClick={undefined}
      type="checkbox" />
  </div>
</li>
`;

exports[`view should render a todo for editing 1`] = `
<input
  autoFocus={true}
  className="edit"
  defaultValue="item 1"
  onBlur={undefined}
  onKeyUp={undefined} />
`;

exports[`view should render a view 1`] = `
<div
  className="todoapp">
  <div
    className="header">
    <h1>
      todos
    </h1>
    <input
      autoFocus={true}
      className="new-todo"
      onKeyUp={[Function]}
      placeholder="What needs to be done?" />
  </div>
  <section
    className="main">
    <input
      className="toggle-all"
      onClick={[Function]}
      type="checkbox" />
    <ul
      className="todo-list" />
  </section>
</div>
`;

exports[`view should render an un-selected filter 1`] = `
<li>
  <a
    className=""
    onClick={[Function]}>
    All
  </a>
</li>
`;
